### DISCUSSION TITLE
Help: How to deal with color specs of 2 components

### DISCUSSION THREAD

**Discussion by mailsnathaniel** [pymupdf/PyMuPDF#1332] [Looking for help] [answered]
[35618400_3.pdf](https://github.com/pymupdf/PyMuPDF/files/7379955/35618400_3.pdf)
Attached PDF file got crashed with the following error message.

Traceback (most recent call last):
  File "pdf_extract.py", line 483, in <module>
    process_pdf(pdfname, opath, doc_counter, 0)
  File "pdf_extract.py", line 422, in process_pdf
    getLineAndRectanges(page, reslist, output_path, pageno)
  File "pdf_extract.py", line 328, in getLineAndRectanges
    fill_opacity=path["opacity"],  # opacity parameters
  File "/root/VENV/mupdf_1_18_6_venv/lib/python3.6/site-packages/fitz/utils.py", line 3752, in finish
    fill_str = ColorCode(fill, "f")  # ensure proper fill string
  File "/root/VENV/mupdf_1_18_6_venv/lib/python3.6/site-packages/fitz/fitz.py", line 2788, in ColorCode
    CheckColor(c)
  File "/root/VENV/mupdf_1_18_6_venv/lib/python3.6/site-packages/fitz/fitz.py", line 2780, in CheckColor
    raise ValueError("need 1, 3 or 4 color components in range 0 to 1")
ValueError: need 1, 3 or 4 color components in range 0 to 1

System Configurations:
 - Operating system: Ubuntu-18.0
 - Python version: 3.6 
 - PyMuPDF version: 1.19.0
For example, the output of `print(sys.version, "\n", sys.platform, "\n", fitz.__doc__)` would be sufficient (for the first two bullets).

**Comment by JorjMcKie** [comment]
Insufficient bug description!
The file looks ok and can be opened with pymupdf no problem.
Also the message says "ValueError" - this is an exception, not a crash.
So what did you do that worked before v1.19.0 and now doesn't anymore?

**Comment by mailsnathaniel** [comment]
I am not saying it worked before. I have processed this document in v1.18 also, it is giving the same error. Kindly please let me know how i can process this document with exception.

Thanks in Advance!

**Comment by JorjMcKie** [comment]
Please do not let me guess - why don't you want to share the failing code.
This exception is very normal if your program is wrong, etc, etc.

**Comment by mailsnathaniel** [comment]
Failing codes are (0.30000001192092896, 0.800000011920929).

**Comment by JorjMcKie** [comment]
We are talking past each other.
I need to see your **_source code_** snippet that causes this exception (again: it is **_not_** a crash). You are obviously trying to output something (text, drawing, annotation, ... **_what?_**), that uses a color specification with 2 components, while PyMuPDF only allows either 1, 3 or 4 components.

I am changing the title and converting the issue to a discussion, category "Looking for Help".

**Comment by mailsnathaniel** [comment]
Hi JorjMcKie, Im trying to access lines and rectangles from fitz and writing it to another file using fitz.open().


<img width="1113" alt="Screenshot 2021-10-21 at 3 50 05 PM" src="https://user-images.githubusercontent.com/64656539/138259482-29b80b3a-655c-4a4e-8825-297c7a5193b2.png">

**Reply by JorjMcKie** [reply]
Ahhh - finally we have something!
Ok, I had a look in your testfile: it contains colorspaces which PyMuPDF does not support yet. These exotic colorspaces come with only 2 color components - which `page.get_drawings()` still extracts correctly, but which are not supported by the `Shape` object,
**_This means, that you cannot redraw such a drawing._** At least not with an equal-looking color.
What you can do as a circumvention:
* check the number of color components in `path["fill"]` / `path["color"]`
* if not supported, take a color of your choice, e.g. `black = (0,)` or `blue = (0, 0, 1)` or whatever.

I will see what I can do in the next version to produce only colors supported by output functions.

**Reply by mailsnathan** [reply]
Sure. Will try this. Thank you very much for your support. 👍

**Reply by JorjMcKie** [reply]
Let me know how it worked.

**Comment by JorjMcKie** [comment, marked as answer]
Your issue has been resolved in the new version.

**Reply by mailsnathan** [reply]
Thanks for new version. Any how i was able to add the exception and it ran without crash.
